File: frontend/food_delivery_flutter/lib/models/user_model.dart
// ============================================================
// FILE: lib/models/user_model.dart
// PURPOSE: User data model for authentication and user profile
// ============================================================

class User {
  final String userId;
  final String email;
  final String name;
  final String? phone;
  final String? address;
  final String? city;
  final String? role; // 'customer', 'restaurant_owner', 'delivery_agent'
  final DateTime createdAt;
  final DateTime updatedAt;
  final int? totalOrders;

  User({
    required this.userId,
    required this.email,
    required this.name,
    this.phone,
    this.address,
    this.city,
    this.role = 'customer',
    required this.createdAt,
    required this.updatedAt,
    this.totalOrders = 0,
  });

  /// Convert User object to JSON (for API requests)
  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'email': email,
      'name': name,
      'phone': phone,
      'address': address,
      'city': city,
      'role': role,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'total_orders': totalOrders,
    };
  }

  /// Create User object from JSON (API responses)
  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      userId: json['user_id'] ?? json['id'] ?? '',
      email: json['email'] ?? '',
      name: json['name'] ?? json['full_name'] ?? '',
      phone: json['phone'],
      address: json['address'],
      city: json['city'],
      role: json['role'] ?? 'customer',
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : DateTime.now(),
      totalOrders: json['total_orders'] ?? 0,
    );
  }

  /// Create a copy of User with updated fields
  User copyWith({
    String? userId,
    String? email,
    String? name,
    String? phone,
    String? address,
    String? city,
    String? role,
    DateTime? createdAt,
    DateTime? updatedAt,
    int? totalOrders,
  }) {
    return User(
      userId: userId ?? this.userId,
      email: email ?? this.email,
      name: name ?? this.name,
      phone: phone ?? this.phone,
      address: address ?? this.address,
      city: city ?? this.city,
      role: role ?? this.role,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      totalOrders: totalOrders ?? this.totalOrders,
    );
  }

  /// Check if user profile is complete
  bool isProfileComplete() {
    return phone != null &&
        address != null &&
        city != null;
  }

  /// Get user's full address
  String getFullAddress() {
    List<String> parts = [];
    if (address != null) parts.add(address!);
    if (city != null) parts.add(city!);
    return parts.join(', ');
  }

  /// Check if user is a customer
  bool isCustomer() => role == 'customer';

  /// Check if user is a restaurant owner
  bool isRestaurantOwner() => role == 'restaurant_owner';

  /// Check if user is a delivery agent
  bool isDeliveryAgent() => role == 'delivery_agent';

  @override
  String toString() {
    return 'User(userId: $userId, email: $email, name: $name, role: $role)';
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is User &&
          runtimeType == other.runtimeType &&
          userId == other.userId &&
          email == other.email;

  @override
  int get hashCode => userId.hashCode ^ email.hashCode;
}


// ============================================================
// USER REGISTRATION MODEL (for signup)
// ============================================================

class UserRegistration {
  final String email;
  final String password;
  final String name;
  final String? phone;

  UserRegistration({
    required this.email,
    required this.password,
    required this.name,
    this.phone,
  });

  Map<String, dynamic> toJson() {
    return {
      'email': email,
      'password': password,
      'name': name,
      'phone': phone,
    };
  }
}


// ============================================================
// USER LOGIN MODEL (for signin)
// ============================================================

class UserLogin {
  final String email;
  final String password;

  UserLogin({
    required this.email,
    required this.password,
  });

  Map<String, dynamic> toJson() {
    return {
      'email': email,
      'password': password,
    };
  }
}


// ============================================================
// LOGIN RESPONSE MODEL
// ============================================================

class LoginResponse {
  final String status;
  final String message;
  final String? token;
  final String? userId;
  final String? email;
  final String? name;
  final User? user;

  LoginResponse({
    required this.status,
    required this.message,
    this.token,
    this.userId,
    this.email,
    this.name,
    this.user,
  });

  bool get isSuccess => status.toLowerCase() == 'success';

  factory LoginResponse.fromJson(Map<String, dynamic> json) {
    return LoginResponse(
      status: json['status'] ?? '',
      message: json['message'] ?? '',
      token: json['token'],
      userId: json['user_id'],
      email: json['email'],
      name: json['name'],
      user: json['user'] != null ? User.fromJson(json['user']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'status': status,
      'message': message,
      'token': token,
      'user_id': userId,
      'email': email,
      'name': name,
      'user': user?.toJson(),
    };
  }
}


// ============================================================
// AUTH STATE (for Provider/Riverpod state management)
// ============================================================

class AuthState {
  final bool isAuthenticated;
  final User? user;
  final String? token;
  final bool isLoading;
  final String? error;

  AuthState({
    this.isAuthenticated = false,
    this.user,
    this.token,
    this.isLoading = false,
    this.error,
  });

  /// Check if user has valid token
  bool get isLoggedIn => isAuthenticated && token != null && user != null;

  /// Create a copy with updated fields
  AuthState copyWith({
    bool? isAuthenticated,
    User? user,
    String? token,
    bool? isLoading,
    String? error,
  }) {
    return AuthState(
      isAuthenticated: isAuthenticated ?? this.isAuthenticated,
      user: user ?? this.user,
      token: token ?? this.token,
      isLoading: isLoading ?? this.isLoading,
      error: error ?? this.error,
    );
  }
}


// ============================================================
// PROFILE UPDATE MODEL
// ============================================================

class ProfileUpdate {
  final String? name;
  final String? phone;
  final String? address;
  final String? city;

  ProfileUpdate({
    this.name,
    this.phone,
    this.address,
    this.city,
  });

  Map<String, dynamic> toJson() {
    return {
      if (name != null) 'name': name,
      if (phone != null) 'phone': phone,
      if (address != null) 'address': address,
      if (city != null) 'city': city,
    };
  }
}

File: backend/models/user_model.py
from datetime import datetime
from bson import ObjectId
from typing import Optional, Dict, Any
 
class User:
    """User model representing a customer in the system"""
    
    def __init__(
        self,
        name: str,
        email: str,
        password_hash: str,
        phone: str,
        role: str = "customer",
        _id: Optional[ObjectId] = None,
        created_at: Optional[datetime] = None,
        updated_at: Optional[datetime] = None,
        address: Optional[str] = None,
        city: Optional[str] = None,
        is_active: bool = True,
        total_orders: int = 0,
    ):
        self._id = _id or ObjectId()
        self.name = name
        self.email = email
        self.password_hash = password_hash
        self.phone = phone
        self.role = role
        self.address = address
        self.city = city
        self.is_active = is_active
        self.total_orders = total_orders
        self.created_at = created_at or datetime.utcnow()
        self.updated_at = updated_at or datetime.utcnow()
 
    def to_dict(self, include_password: bool = False) -> Dict[str, Any]:
        """Convert user object to dictionary for MongoDB"""
        data = {
            "_id": self._id,
            "name": self.name,
            "email": self.email,
            "phone": self.phone,
            "role": self.role,
            "address": self.address,
            "city": self.city,
            "is_active": self.is_active,
            "total_orders": self.total_orders,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
        }
        
        if include_password:
            data["password"] = self.password_hash
        
        return data
 
    def to_json(self, include_password: bool = False) -> Dict[str, Any]:
        """Convert user to JSON-serializable format for API responses"""
        data = {
            "user_id": str(self._id),
            "name": self.name,
            "email": self.email,
            "phone": self.phone,
            "role": self.role,
            "address": self.address,
            "city": self.city,
            "is_active": self.is_active,
            "total_orders": self.total_orders,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }
        
        if include_password:
            data["password"] = self.password_hash
        
        return data
 
    @staticmethod
    def from_dict(data: Dict[str, Any]) -> "User":
        """Create User object from dictionary (MongoDB document)"""
        return User(
            name=data.get("name", ""),
            email=data.get("email", ""),
            password_hash=data.get("password", ""),
            phone=data.get("phone", ""),
            role=data.get("role", "customer"),
            _id=data.get("_id"),
            created_at=data.get("created_at"),
            updated_at=data.get("updated_at"),
            address=data.get("address"),
            city=data.get("city"),
            is_active=data.get("is_active", True),
            total_orders=data.get("total_orders", 0),
        )
 
    @staticmethod
    def from_json(data: Dict[str, Any]) -> "User":
        """Create User object from JSON (API requests)"""
        return User(
            name=data.get("name", ""),
            email=data.get("email", ""),
            password_hash=data.get("password", ""),
            phone=data.get("phone", ""),
            role=data.get("role", "customer"),
            address=data.get("address"),
            city=data.get("city"),
        )
 
    def update(self, **kwargs) -> None:
        """Update user fields and set updated_at timestamp"""
        allowed_fields = ["name", "phone", "address", "city", "is_active"]
        for key, value in kwargs.items():
            if key in allowed_fields and value is not None:
                setattr(self, key, value)
        self.updated_at = datetime.utcnow()
 
    def is_profile_complete(self) -> bool:
        """Check if user has completed their profile"""
        return all([
            self.name and len(self.name) >= 2,
            self.email,
            self.phone,
            self.address,
            self.city,
        ])
 
    def get_full_address(self) -> str:
        """Get formatted full address"""
        parts = []
        if self.address:
            parts.append(self.address)
        if self.city:
            parts.append(self.city)
        return ", ".join(parts) if parts else "Not provided"
 
    def __repr__(self) -> str:
        return f"<User(id={self._id}, email={self.email}, name={self.name})>"
 
    def __eq__(self, other) -> bool:
        if not isinstance(other, User):
            return False
        return self._id == other._id and self.email == other.email
 
    def __hash__(self) -> int:
        return hash(str(self._id))